warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java

```diff
@@ -0,0 +1,291 @@
+package datawave.query.postprocessing.tf;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.Multimap;
+import datawave.query.attributes.Attribute;
+import datawave.query.attributes.Attributes;
+import datawave.query.attributes.Document;
+import datawave.query.attributes.PreNormalizedAttribute;
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.accumulo.core.data.Key;
+import org.apache.commons.jexl2.parser.ASTIdentifier;
+import org.apache.commons.jexl2.parser.ASTNumberLiteral;
+import org.apache.commons.jexl2.parser.ASTOrNode;
+import org.apache.commons.jexl2.parser.ASTUnaryMinusNode;
+import org.apache.commons.jexl2.parser.JexlNode;
+import org.apache.log4j.Logger;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_ADJACENT_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_PHRASE_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_SCORED_PHRASE_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_WITHIN_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME;
+
+/**
+ * A function that intersects document keys prior to fetching term offsets.
+ * <p>
+ * This only works on positive phrase queries -- that is, no guarantee is made for negated phrases
+ */
+public class DocumentKeysFunction {
+    
+    private static final Logger log = Logger.getLogger(DocumentKeysFunction.class);
+    
+    private final Multimap<String,ContentFunction> contentFunctions = LinkedListMultimap.create();
+    
+    /**
+     * @param config
+     *            a {@link TermFrequencyConfig}
+     */
+    public DocumentKeysFunction(TermFrequencyConfig config) {
+        populateContentFunctions(config.getScript());
+    }
+    
+    protected void populateContentFunctions(JexlNode node) {
+        
+        Multimap<String,Function> functions = TermOffsetPopulator.getContentFunctions(node);
+        for (String key : functions.keySet()) {
+            Collection<Function> coll = functions.get(key);
+            for (Function f : coll) {
+                ContentFunction contentFunction = transformFunction(f);
+                contentFunctions.put(contentFunction.getField(), contentFunction);
+            }
+        }
+    }
+    
+    private ContentFunction transformFunction(Function function) {
```

Comment:
This also seems like a weird place for this functionality.

Reply:
Fair